### PR TITLE
Don't load empty cells; use empirical worksheet dimensions

### PR DESCRIPTION
**Pull request by jennybc**
There can be cells that look empty to the human eye, but that are styled, e.g. someone applied a custom number format to the cell at some point. I propose we do not load those anymore. There's nothing to learn from their XML.

If such a cell still falls into the target rectangle, it will simply be `NA` in the output (ok, won't be totally true until we stop dropping unnamed, empty columns #157).

If such a cell falls outside the target rectangle, this PR prevents the creation of trailing row(s) or column(s) consisting entirely of `NA`.

In addition to rows/columns of `NA`, this has been a puzzling source of errors about column names and types not being compatible.

To fully fix this, you also can't trust a worksheet's declared dimensions, because these cells "count". Therefore I propose we always compute dimension ourselves.

I haven't dealt with xls yet. I can leave this open and work on that. Or we could resolve this for xlsx and I'd open an issue to fix it for xls.

This branch builds off #247. I'm assuming it will be merged first.

### CODE REVIEW

File: src/XlsxWorkSheet.h [hunk 1]
```diff
@@ -296,16 +267,20 @@ class XlsxWorkSheet {
       int j = 0;
       for (rapidxml::xml_node<>* cell = row->first_node("c");
            cell; cell = cell->next_sibling("c")) {
-        XlsxCell xcell(cell, i, j);
-        cells_.push_back(xcell);
+        // don't load a cell with no child nodes, e.g. it only has style
+        rapidxml::xml_node<>* first_child = cell->first_node(0);
```
Comment (hadley) [inline src/XlsxWorkSheet.h]: I think the check here probably eliminates a check somewhere else.

File: tests/testthat/test-empty.R [hunk 2]
```diff
@@ -28,3 +28,31 @@ test_that("non-empty sheets act that way if we skip past everything", {
...
+test_that("empty cells with a style are not loaded", {
+  ## what's important about this sheet?
+  ## contains empty cells with a custom format
+  ## therefore they appear in the xml and have a style attribute
+  ## where are they?
+  ## in the embedded empty columns, w/ and w/o a name
+  ## in a trailing empty column
+  ## in some trailing rows
+  out <- read_excel(test_sheet("style-only-cells.xlsx"))
+  expect_equal(
```
Comment (hadley) [inline tests/testthat/test-empty.R]: I think code like this is slightly cleaner if you define the tibble outside the test

File: tests/testthat/test-empty.R [hunk 3]
```diff
@@ -28,3 +28,31 @@ test_that("non-empty sheets act that way if we skip past everything", {
...
+                   var3 = c("aa", "bb", "cc"),
+                   var5 = c(1, 2, 3))
+  )
+  skip("revisit this when dust settles re: treatment of empty columns")
+})
+
+test_that("user-supplied column names play nicely with empty columns", {
+  skip("waiting for dust to settle re: treatment of empty columns")
+  ## do stuff like this:
+  out <- read_excel(test_sheet("style-only-cells.xlsx"),
```
Comment (hadley) [inline tests/testthat/test-empty.R]: House style is

```R
read_excel(
  test_sheet("style-only-cells.xlsx"),
  col_names = LETTERS[1:4]
)
```

File: src/XlsxWorkSheet.h (hunk 1 above)
Comment (jennybc) [inline src/XlsxWorkSheet.h]: There's not an obvious one, but I'll keep my eyes peeled.

Styled empty cells were a major source of "cell lacks the t attribute" here, which is why they always lead to numeric `NA` columns. But it's not clear they are the only source? When I get to column types, I'll try to settle this definitively.

https://github.com/tidyverse/readxl/blob/a78440d4f4f4fa62f8b89fe5296a262e9cb24180/src/XlsxCell.h#L122-L124

Comment (jennybc) [inline src/XlsxWorkSheet.h]: Similar situation with this check for a cell having a `v` node. But I'm not sure if these cells are the only source of this either?

https://github.com/tidyverse/readxl/blob/a78440d4f4f4fa62f8b89fe5296a262e9cb24180/src/XlsxCell.h#L102-L104

File: tests/testthat/test-empty.R (hunk 2 above)
Comment (jennybc) [inline tests/testthat/test-empty.R]: done

File: tests/testthat/test-empty.R (hunk 3 above)
Comment (jennybc) [inline tests/testthat/test-empty.R]: done

**Pull request merged**